src/loss/WLS/WLS.jl

```diff
@@ -58,269 +59,98 @@ function SemWLS(;
+            )
         end
     else
+        isnothing(wls_weight_matrix_mean) ||
+            @warn "Ignoring wls_weight_matrix_mean since meanstructure is disabled"
         wls_weight_matrix_mean = nothing
     end
+    HE = approximate_hessian ? ApproximateHessian : ExactHessian
 
-    return SemWLS(
-        wls_weight_matrix,
-        s,
-        approximate_hessian,
-        wls_weight_matrix_mean,
-        Val(meanstructure),
-    )
+    return SemWLS{HE}(wls_weight_matrix, s, wls_weight_matrix_mean)
 end
 
 ############################################################################
 ### methods
 ############################################################################
 
-objective!(semwls::SemWLS, par, model::AbstractSemSingle) =
-    objective!(semwls::SemWLS, par, model, semwls.has_meanstructure)
-gradient!(semwls::SemWLS, par, model::AbstractSemSingle) =
-    gradient!(semwls::SemWLS, par, model, semwls.has_meanstructure)
-hessian!(semwls::SemWLS, par, model::AbstractSemSingle) =
-    hessian!(semwls::SemWLS, par, model, semwls.has_meanstructure)
-
-objective_gradient!(semwls::SemWLS, par, model::AbstractSemSingle) =
-    objective_gradient!(semwls::SemWLS, par, model, semwls.has_meanstructure)
-objective_hessian!(semwls::SemWLS, par, model::AbstractSemSingle) =
-    objective_hessian!(semwls::SemWLS, par, model, semwls.has_meanstructure)
-gradient_hessian!(semwls::SemWLS, par, model::AbstractSemSingle) =
-    gradient_hessian!(semwls::SemWLS, par, model, semwls.has_meanstructure)
-
-objective_gradient_hessian!(semwls::SemWLS, par, model::AbstractSemSingle) =
-    objective_gradient_hessian!(semwls::SemWLS, par, model, semwls.has_meanstructure)
-
-function objective!(
+function evaluate!(
+    objective,
+    gradient,
+    hessian,
     semwls::SemWLS,
-    par,
+    implied::SemImplySymbolic,
     model::AbstractSemSingle,
-    has_meanstructure::Val{T},
-) where {T}
-    let σ = Σ(imply(model)),
-        μ = μ(imply(model)),
-        σₒ = semwls.σₒ,
-        μₒ = obs_mean(observed(model)),
-        V = semwls.V,
-        V_μ = semwls.V_μ,
-
-        σ₋ = σₒ - σ
-
-        if T
-            μ₋ = μₒ - μ
-            return dot(σ₋, V, σ₋) + dot(μ₋, V_μ, μ₋)
-        else
-            return dot(σ₋, V, σ₋)
-        end
-    end
-end
-
-function gradient!(
-    semwls::SemWLS,
     par,
-    model::AbstractSemSingle,
-    has_meanstructure::Val{T},
-) where {T}
-    let σ = Σ(imply(model)),
-        μ = μ(imply(model)),
-        σₒ = semwls.σₒ,
-        μₒ = obs_mean(observed(model)),
-        V = semwls.V,
-        V_μ = semwls.V_μ,
-        ∇σ = ∇Σ(imply(model)),
-        ∇μ = ∇μ(imply(model))
-
-        σ₋ = σₒ - σ
-
-        if T
-            μ₋ = μₒ - μ
-            return -2 * (σ₋' * V * ∇σ + μ₋' * V_μ * ∇μ)'
-        else
-            return -2 * (σ₋' * V * ∇σ)'
-        end
+)
+    if !isnothing(hessian) && (MeanStructure(implied) === HasMeanStructure)
```

Comment:
```suggestion
    if !isnothing(hessian) && (MeanStructure(implied) == HasMeanStructure)
```

Probably does not make a difference, but I think in general, you don't need to compare the equality of two types with `===` but can just use `==`, as there is a method for that defined in Base julia:

```
julia> @which Float64 == Float64
==(T::Type, S::Type)
     @ Base operators.jl:250
```

Reply:
You are correct. I just tend to use `===` to indicate in the code that this is a compile-time traits check. Also `===` is guaranteed to be strict and fast "egality" comparison without potential method overloading complications.
So the question is whether you would prefer that in the SEM.jl code `===` is used to highlight compile-time checks or you prefer using `==` universally.
I can adjust the PR accordingly.

Reply:
I think it makes total sense to have `===` highlighting compile time checks, so I would leave it as is 👍🏽